include/micm/util/sparse_matrix.hpp

```diff
@@ -0,0 +1,205 @@
+// Copyright (C) 2023 National Center for Atmospheric Research,
+//
+// SPDX-License-Identifier: Apache-2.0
+#pragma once
+
+#include <cassert>
+#include <stdexcept>
+#include <utility>
+#include <vector>
+
+namespace micm
+{
+
+  template<class T>
+  class SparseMatrixBuilder;
+
+  /// @brief A sparse block-diagonal 2D matrix class with contiguous memory
+  ///
+  /// Each block sub-matrix is square and has the same structure of non-zero elements
+  ///
+  /// Sparse matrix data structure follows the Compressed Sparse Row (CSR) pattern
+  template<class T>
+  class SparseMatrix
+  {
+    std::size_t number_of_blocks_;        // Number of block sub-matrices in the overall matrix
+    std::vector<T> data_;                 // Value of each non-zero matrix element
+    std::vector<std::size_t> row_ids_;    // Row indices of each non-zero element in a block
+    std::vector<std::size_t> row_start_;  // Index in data_ and row_ids_ of the start of each column in a block
+
+    friend class SparseMatrixBuilder<T>;
+    friend class ProxyRow;
+    friend class Proxy;
+
+    class Proxy
+    {
+      SparseMatrix& matrix_;
+      std::size_t block_id_;
+      std::size_t row_id_;
+
+     public:
+      Proxy(SparseMatrix& matrix, std::size_t block_id, std::size_t row_id)
+          : matrix_(matrix),
+            block_id_(block_id),
+            row_id_(row_id)
+      {
+      }
+
+      std::size_t size() const
+      {
+        return matrix_.row_start_.size() - 1;
+      }
+
+      T& operator[](std::size_t y)
+      {
+        return matrix_.data_[matrix_.VectorIndex(block_id_, row_id_, y)];
+      }
+    };
+
+    class ProxyRow
+    {
+      SparseMatrix& matrix_;
+      std::size_t block_id_;
+
+     public:
+      ProxyRow(SparseMatrix& matrix, std::size_t block_id)
+          : matrix_(matrix),
+            block_id_(block_id)
+      {
+      }
+
+      std::size_t size() const
+      {
+        return matrix_.row_start_.size() - 1;
+      }
+
+      Proxy operator[](std::size_t x)
+      {
+        return Proxy(matrix_, block_id_, x);
+      }
+    };
+
+   public:
+    static SparseMatrixBuilder<T> create(std::size_t block_size)
+    {
+      return SparseMatrixBuilder<T>{ block_size };
+    }
+
+    SparseMatrix(SparseMatrixBuilder<T>& builder)
+        : number_of_blocks_(builder.number_of_blocks_),
+          data_(builder.NumberOfElements(), builder.initial_value_),
+          row_ids_(builder.RowIdsVector()),
+          row_start_(builder.RowStartVector())
+    {
+    }
+
+    std::vector<T>& AsVector()
+    {
+      return data_;
+    }
+
+    std::size_t VectorIndex(std::size_t block, std::size_t row, std::size_t column) const
+    {
+      if (row >= row_start_.size() - 1 || column >= row_start_.size() - 1 || block >= number_of_blocks_)
+        throw std::invalid_argument("SparseMatrix element out of range");
```

Comment:
Normally I would say we shouldn't throw, but in this case it's an irrecoverable error, it seems like. The only other option I can think of is like what I did for the SolverResult, but I'm not really sure that would win us anything. I think this is fine.